anchore_engine/clients/grype_wrapper.py

```diff
@@ -10,322 +10,433 @@
+    VULNERABILITY_FILE_NAME = "vulnerability.db"
+    METADATA_FILE_NAME = "metadata.json"
+    ARCHIVE_FILE_NOT_FOUND_ERROR_MESSAGE = "New grype_db archive file not found"
+    MISSING_GRYPE_DB_DIR_ERROR_MESSAGE = (
+        "Cannot access missing grype_db dir. Reinitialize grype_db."
     )
-    os.remove(grype_db_archive_local_file_location)
-
-
-def _move_and_open_grype_db_archive(
-    grype_db_archive_local_file_location: str, version_name: str
-) -> str:
-    """
-    This function moves a tarball containing the latest grype db from a location on the local file system
-    into the configured grype db dir. It then extracts all files in the tarball and removes the then-unneeded
-    archive file.
-    """
-    # Get the location to copy the archive to
-    local_db_dir = _get_default_grype_db_dir_from_config()
-
-    # Move the archive
-    grype_db_archive_copied_file_location = _move_grype_db_archive(
-        grype_db_archive_local_file_location, local_db_dir
+    MISSING_GRYPE_DB_SESSION_MAKER_ERROR_MESSAGE = (
+        "Cannot access missing grype_db session maker. Reinitialize grype_db."
     )
 
-    # Unpack the archive
-    latest_grype_db_dir = _open_grype_db_archive(
-        grype_db_archive_copied_file_location, local_db_dir, version_name
-    )
+    def __new__(cls):
+        if cls._grype_wrapper_instance is None:
+            cls._grype_wrapper_instance = super(GrypeWrapperSingleton, cls).__new__(cls)
+            cls._grype_db_dir_internal = None
+            cls._grype_db_session_maker_internal = None
+            cls._grype_db_lock = rwlock.RWLockWrite()
+        return cls._grype_wrapper_instance
+
+    @classmethod
+    def get_instance(cls):
+        """
+        Returns the singleton instance of this class.
+        """
+        return GrypeWrapperSingleton()
+
+    @classmethod
+    def _get_test_instance(cls):
+        """
+        Returns a new instance of this class. This method is not intended for use outside of tests.
+        """
+        cls._grype_wrapper_instance = None
+        return GrypeWrapperSingleton()
+
+    @property
+    def _grype_db_dir(self):
+        if self._grype_db_dir_internal is None:
+            raise ValueError(self.MISSING_GRYPE_DB_DIR_ERROR_MESSAGE)
+        else:
+            return self._grype_db_dir_internal
+
+    @_grype_db_dir.setter
+    def _grype_db_dir(self, _grype_db_dir_internal):
+        self._grype_db_dir_internal = _grype_db_dir_internal
+
+    @property
+    def _grype_db_session_maker(self):
+        if self._grype_db_session_maker_internal is None:
+            raise ValueError(self.MISSING_GRYPE_DB_SESSION_MAKER_ERROR_MESSAGE)
+        else:
+            return self._grype_db_session_maker_internal
+
+    @_grype_db_session_maker.setter
+    def _grype_db_session_maker(self, _grype_db_session_maker_internal):
+        self._grype_db_session_maker_internal = _grype_db_session_maker_internal
+
+    @contextmanager
+    def read_lock_access(self):
```

Comment:
This is a nice enhancement. If you had implemented these as a class, it would give you the option of parametrizing the context manager through the constructor for the class, which would mean that you could modify the timeouts for different situations. But if you don't need such a granular level of control, then this is fine.